BaseCryptOnOneCrypto Support (#1570)
Microsoft/mu_basecore: CryptoPkg/Library/BaseCryptLibOnOneCrypto/readme.md

3. Package the binaries with the updated version number
4. Create a GitHub release on the mu_crypto_release repository

For detailed information on building and publishing the OneCrypto binary driver,
visit the [mu_crypto_release](https://github.com/microsoft/mu_crypto_release) repository.

---

### Step 7: Update the External Dependency (extdep)

Update the external dependency descriptor to pull in the newly built OneCrypto binary with your new functions.

**Location:** `*_ext_dep.json` or similar dependency configuration file

**Example:**

```json
{
  "scope": "global",
  "type": "web",
  "name": "edk2-onecrypto-driver-bin",
  "source": "https://github.com/microsoft/mu_crypto_release/releases/download/v1.1.0/Mu_CryptoBin_v1_0_0.zip",
  "version": "1.0.0",
  "sha256": "b62686b7dc6331c22ed48c48af76ee4feb525545e866feae5b32548d75eab28a",
  "compression_type": "zip",
  "internal_path": "/",
  "flags": ["set_build_var"],
  "var_name": "BLD_*_ONE_CRYPTO_PATH"
}
```

**Process:**

1. Build the OneCrypto binary driver with your new implementation
2. Publish the package to your github repository
3. Update the version number in `ext_dep.json`

> 4. update the sha256 of the expected packaged crypto

Reviewer: ```suggestion
4. update the sha256 of the expected packaged crypto in `ext_dep.json"

```

> 5. Test that the build system correctly downloads and extracts the new binary

Reviewer: mention stuart_update for step 5?

Author: Edk2 doesn't use stuart right? How do ext deps work for them?

# BaseCryptLibOnOneCrypto

## Overview

> BaseCryptLibOnOneCrypto provides the BaseCryptLib and TlsLib APIs by calling the OneCrypto Protocol/PPI.

Reviewer 2: In a lot of places you mention the PPI. There is no PPI today. I'm fine to have this be aspirational, but in some places you have the DXE version mentioning the PPI, e.g., so I would remove those cases.

**Platform code must never call the OneCrypto protocol directly.** Always use this library to ensure version safety,
type safety, and robust error handling.

### Why Not Call the Protocol Directly?

Direct protocol usage is unsafe: protocol structure and function signatures may change, breaking binary compatibility
and introducing security risks. This library provides:

- **Version validation** (at runtime, using protocol major/minor)
- **Type safety** (strong C API)
- **Consistent error handling**
- **Stable API** (insulates platform code from protocol changes)
- **Automatic initialization and protocol lookup**

## Usage

### INF Example

```inf
[LibraryClasses]

>   BaseCryptLib|CryptoPkg/Library/BaseCryptLibOnOneCrypto/BaseCryptLibOnOneCrypto.inf

Reviewer 2: Funnily, this is what I'm suggesting you move to, but does not currently exist :)

Author: Fixing :)

        class GenericLibrary:::module {
            <<Library Interface>>
            Any library that needs crypto
        }

    }

    OneCryptoLoaderSupvMm --> OneCryptoBinSupvMm : Loads Phase Agnostic Code
    OneCryptoLoaderDxe --> OneCryptoBinSupvMm : Loads Phase Agnostic Code

    OneCryptoBinSupvMm ..> OneCryptoLoaderSupvMm : Provides Crypto
    OneCryptoBinSupvMm ..> OneCryptoLoaderDxe : Provides Crypto

    OneCryptoLoaderSupvMm ..|> OneCrypto : Publishes
    OneCryptoLoaderDxe ..|> OneCrypto : Publishes

    BaseCryptOnOneCrypto ..|> OneCrypto : Locates

    BaseCryptOnOneCrypto <.. BaseCryptLib : Backed By
    BaseCryptOnOneCrypto <.. TlsLib : Backed By

    GenericLibrary ..> BaseCryptLib : Uses
    GenericDriver ..> BaseCryptLib : Uses
    GenericLibrary ..> TlsLib : Uses
    GenericDriver ..> HashApiLib : Uses

    classDef driver fill:#187286,stroke:#333,stroke-width:2px
    classDef protocol fill:#d63230,stroke:#333,stroke-width:2px
    classDef library fill:#1c77c3,stroke:#333,stroke-width:2px
    classDef module fill:#ff8c00,stroke:#333,stroke-width:2px

```

**Architecture Components:**

- **Platform Code**: Uses standard BaseCryptLib/TlsLib/HashApiLib APIs

> - **OneCrypto Loaders**: Contain enough logic to find and load the phase agnostic code

Reviewer 2: ```suggestion
- **OneCrypto Loaders**: Finds and loads the phase agnostic code
```

  $(ONE_CRYPTO_PATH)/OneCryptoBin/OneCryptoBinSupvMm.inf

  # Or for traditional Standalone MM:
  # $(ONE_CRYPTO_PATH)/OneCryptoLoaders/OneCryptoLoaderStandaloneMm.inf
  # $(ONE_CRYPTO_PATH)/OneCryptoBin/OneCryptoBinStandaloneMm.inf
```

And include them in your FDF:

```fdf
INF $(ONE_CRYPTO_PATH)/OneCryptoLoaders/OneCryptoLoaderDxe.inf
INF $(ONE_CRYPTO_PATH)/OneCryptoLoaders/OneCryptoLoaderSupvMm.inf
INF $(ONE_CRYPTO_PATH)/OneCryptoBin/OneCryptoBinSupvMm.inf
```

### In Your C Code

Simply include and use the standard BaseCryptLib API:

```c
#include <Library/BaseCryptLib.h>

EFI_STATUS
MyFunction (
  VOID
  )
{
  UINT8   Hash[32];
  UINTN   HashSize;
  VOID    *Sha256Context;

  // Initialize SHA256 context
  Sha256Context = Sha256New();
  if (Sha256Context == NULL) {
    return EFI_OUT_OF_RESOURCES;
  }

  // Use standard BaseCryptLib functions
  if (!Sha256Init(Sha256Context)) {
    Sha256Free(Sha256Context);
    return EFI_DEVICE_ERROR;
  }

  // The library handles all protocol interactions internally
  Sha256Update(Sha256Context, Data, DataSize);
  Sha256Final(Sha256Context, Hash);
  Sha256Free(Sha256Context);

  return EFI_SUCCESS;
}

> ```

Reviewer 2: This is all duplicate

## Adding New Cryptographic Functions (Detailed Guide)

> This section provides a comprehensive guide for adding new cryptographic functions to the OneCrypto platform.

Reviewer 2: ```suggestion
This section provides a comprehensive guide for adding new cryptographic functions to the OneCrypto protocol.
```

> This section provides a comprehensive guide for adding new cryptographic functions to the OneCrypto platform.

Reviewer 2: It seems like this would better belong in mu_crypto_release and this readme points to that (which doesn't exist yet, so this can have a todo to point to it)

Author: I think Sean wanted this here as well, but I'm also happy to make doc changes as needed

implementations and keeps the codebase organized.

**Example:**

```c
/**
  Allocates and initializes one CRYPTO context for subsequent operations.

  @return  NULL  This interface is not supported.
**/
VOID *
EFIAPI
GenericCryptoNew (
  VOID
  )
{
  ASSERT (FALSE);
  return NULL;
}

/**
  Performs a cryptographic operation on a data buffer of the specified size.

  @param[in]   Context      Pointer to the CRYPTO context.
  @param[in]   InputData    Pointer to the buffer containing the data to be processed.
  @param[in]   InputSize    Size of InputData buffer in bytes.
  @param[out]  OutputData   Pointer to a buffer that receives the output.
  @param[out]  OutputSize   Pointer to the size of the output buffer.

  @retval FALSE  This interface is not supported.
**/
BOOLEAN
EFIAPI
GenericCryptoProcess (
  IN   VOID         *Context,
  IN   CONST VOID   *InputData,
  IN   UINTN        InputSize,
  OUT  UINT8        *OutputData,
  OUT  UINTN        *OutputSize
  )
{
  ASSERT (FALSE);
  return FALSE;
}
```

**Why This Matters:**

- Allows modules to compile without linking real crypto implementations
- Enables CI builds that don't require binary dependencies

> - The `ASSERT(FALSE)` catches accidental usage in debug builds

Reviewer 2: I think this section is mostly duplicative and can be dropped.

- `Hash` - Hashing functions (SHA, MD5, SM3)
- `HMAC` - HMAC functions
- `AES` - AES encryption/decryption
- `BN` - Big number operations
- `HKDF` - Key derivation functions
- `PKCS` - Public key cryptography standards
- `DH` - Diffie-Hellman key exchange
- `EC` - Elliptic curve cryptography
- `RSA` - RSA public key operations
- `X509` - Certificate handling
- `Random` - Random number generation
- `Tls` - TLS/SSL operations
- `Timestamp` - RFC3161 timestamp verification
- `Info` - Library information

Choose the most appropriate group for your function. If adding an entirely new category of cryptographic functions, you
may need to define a new `@defgroup` at the top of the section where you add your functions.

**Critical Rules:**

- **Always add to the end** of the structure (maintains binary compatibility)
- **Increment Minor version** for new functions (Major version changes are breaking changes)
- **Never reorder or remove existing function pointers**
- Document which version introduced each function

---

### Step 4: Update BaseCryptLibOnOneCrypto to Call the Protocol

Implement the BaseCryptLib function by calling the protocol using the `CALL_CRYPTO_SERVICE` macro.

> **Location:** `CryptoPkg/Library/BaseCryptLibOnOneCrypto/CryptoService.c` (or appropriate file)

Reviewer 2: This file doesn't exist. With my suggestion would be in OneCryptoLibOnOneCrypto.c

  1,                                              // Minimum Major version required
  1                                               // Minimum Minor version required
);
```

**Why This Is Safe:**

- The macro checks protocol version at runtime
- If the binary driver is too old (< v1.1), it returns the error value (`FALSE` or `NULL`)
- Platform code that uses newer functions requires a newer binary driver
- Older binary drivers continue to work with platform code that only uses older functions

---

### Step 5: Implement the Function in the OneCrypto Binary Driver

Implement the actual cryptographic functionality in the OneCrypto binary driver and update the protocol structure initialization.

**Location:** `OneCryptoBin/Driver/OneCryptoDriver.c` (or appropriate source file in the binary driver)

**Example Implementation:**

```c
/**
  Allocates and initializes one CRYPTO context for subsequent operations.

  @return  Pointer to the CRYPTO context that has been initialized.
           If the allocations fails, GenericCryptoNew() returns NULL.
**/
VOID *
EFIAPI
CryptoServiceGenericCryptoNew (
  VOID
  )
{

>   return GenericCryptoNew();  // Calls underlying OpenSSL/MbedTLS implementation

Reviewer 2: ```suggestion
  return GenericCryptoNew();  // Calls underlying crypto provider implementation
```

}

/**
  Performs a cryptographic operation on a data buffer of the specified size.

  @param[in]   Context      Pointer to the CRYPTO context.
  @param[in]   InputData    Pointer to the buffer containing the data to be processed.
  @param[in]   InputSize    Size of InputData buffer in bytes.
  @param[out]  OutputData   Pointer to a buffer that receives the output.
  @param[out]  OutputSize   Pointer to the size of the output buffer.

  @retval TRUE   Cryptographic operation succeeded.
  @retval FALSE  Cryptographic operation failed.
**/
BOOLEAN
EFIAPI
CryptoServiceGenericCryptoProcess (
  IN   VOID         *Context,
  IN   CONST VOID   *InputData,
  IN   UINTN        InputSize,
  OUT  UINT8        *OutputData,
  OUT  UINTN        *OutputSize
  )
{
  // Input validation
  if (Context == NULL || OutputData == NULL || OutputSize == NULL) {
    return FALSE;
  }

>   if ((InputData == NULL) && (InputSize != 0)) {

Reviewer 2: I would finish this function out.

> // In OneCryptoBin.c - Protocol instance initialization

Reviewer 2: This is confusing to have here in the middle of a function in a different file.

 CryptoProtocol->CryptoServiceGenericCryptoNew = CryptoServiceGenericCryptoNew;
 CryptoProtocol->CryptoServiceGenericCryptoProcess = CryptoServiceGenericCryptoProcess;
```

**Implementation Notes:**

- Prefix implementation functions with `CryptoService` to distinguish from library calls
- Perform input validation before calling underlying crypto libraries
- Handle errors gracefully and return appropriate status codes

> - Consider hardware acceleration if available

Reviewer 2: This isn't really up to OneCrypto, this is up to the underlying crypto provider. I think you can remove this statement.

---

> ### Step 6: Update the External Dependency (extdep)

Reviewer 2: Step 6 should be build and publish, then update extdep

Update the external dependency descriptor to pull in the newly built OneCrypto binary with your new functions.

**Location:** `*_ext_dep.json` or similar dependency configuration file

**Example:**

```json
{
  "scope": "global",
  "type": "web",
  "name": "edk2-onecrypto-driver-bin",
  "source": "https://github.com/microsoft/mu_crypto_release/releases/download/v1.1.0/Mu_CryptoBin_v1_0_0.zip",
  "version": "1.0.0",
  "sha256": "b62686b7dc6331c22ed48c48af76ee4feb525545e866feae5b32548d75eab28a",
  "compression_type": "zip",
  "internal_path": "/",
  "flags": ["set_build_var"],
  "var_name": "BLD_*_ONE_CRYPTO_PATH"
}
```

**Process:**

1. Build the OneCrypto binary driver with your new implementation
2. Publish the package to your github repository
3. Update the version number in `ext_dep.json`
4. update the sha256 of the expected packaged crypto
5. Test that the build system correctly downloads and extracts the new binary

**Testing the Update:**

```powershell
# Clean up the old dependencies

> python PlatformBuild.py --clean # Or just delete the ext dep binary folder

Reviewer 2: This isn't needed, stuart will handle this.

In general here, I would use the stuart commands directly, not all PlatformBuild.py's have this wrapper built in.

# Update the dependencies
python PlatformBuild.py --update

# Verify the correct version is present
ls $(ONE_CRYPTO_PATH)
```

---

> ### Step 7: Add Tests and Update Documentation

Reviewer 2: In order for tests to work, you also have to update BaseCryptLib itself so that you can test properly, right? Or these are runtime tests?

Author: The assumption is you did that in Step 1

- [ ] **Step 4.1:** Implementation added to `BaseCryptLibOnOneCrypto` using `CALL_CRYPTO_SERVICE`
- [ ] **Step 4.2:** Correct version numbers specified in macro (matching protocol version)
- [ ] **Step 5.1:** Function implemented in OneCrypto binary driver
- [ ] **Step 5.2:** Protocol structure initialization updated with new function pointers
- [ ] **Step 6.1:** External dependency updated to new binary version
- [ ] **Step 6.2:** Dependency correctly downloads and extracts
- [ ] **Step 7.1:** Unit tests written and passing
- [ ] **Step 7.2:** API documentation updated
- [ ] **Step 7.3:** Release notes created
- [ ] **Step 7.4:** Integration examples provided

### Version Compatibility Matrix

| Platform Code Uses | Binary Driver Version | Result |
|-------------------|-----------------------|--------|
| v1.0 functions | v1.0 binary | ✅ Works |
| v1.0 functions | v1.1 binary | ✅ Works (backward compatible) |
| v1.1 functions | v1.1 binary | ✅ Works |
| v1.1 functions | v1.0 binary | ⚠️ New functions return error, old functions work |
| v2.0 functions | v1.1 binary | ❌ Fails (major version mismatch) |

This ensures smooth upgrades and prevents breaking existing platforms.

> ## The `CALL_CRYPTO_SERVICE` Macro

Reviewer 2: You describe this above, only keep one section and again with my suggestion this goes away.
